### ISSUE TITLE
Use CMake

### ISSUE THREAD

**Issue by vtraag** [igraph/igraph#1185]
The `igraph` library could switch to `CMake` for managing the build process, also in order to support correct and easier compilation for MSVC. This would help address issues #969, #909, #1169 on correctly compiling in Windows.

As suggested by @szhorvat we could add `CMake` for 0.8.0 and switch to it completely for 0.9.0. We can discuss and track our further discussion here.

**Comment by vtraag** [comment]
I have been trying to take a look at `CMake`. It will take me quite some time to familiarize myself with `CMake` and the build requirements and exact flow of the build process for `igraph`. Unless there is someone who is already relatively familiar with `CMake` and knows how to translate things from `automake` into `CMake`, I think it will require too much time to set it up now. I would therefore propose to postpone it until 0.9.0.

In addition, I think it may actually be more complicated to support two build systems simultaneously, so it might be easier to switch completely to `CMake` for 0.9.0 instead of holding on to both `automake` and `CMake`.

What are your thoughts?

**Comment by ntamas** [comment]
I was planning to take a look at adapting igraph's build system to CMake, but unfortunately I did not have time to do so in the last few weeks (and probably won't have time for it until September either). Maintaining two build systems in parallel probably doesn't make sense in the long run, so let's aim for replacing `make` with `cmake` entirely by 0.9.0.

**Comment by iosonofabio** [comment]
+1 for switching to cmake or staying with autotools as opposed to maintaining two systems :-)

**Comment by ntamas** [comment]
Progress report: CMake support now works in the [feature/cmake-build](https://github.com/igraph/igraph/tree/feature/cmake-build) branch. More precisely, one can build a static _or_ a shared igraph library with CMake and run all the tests.

Things that are not supported yet:

* building the documentation

* building a static and a shared library from the same configuration (and I'm not sure whether we'll support it, for two reasons: 1) it goes against the CMake philosophy where there is a single setting that controls whether you want a static or a shared lib at the end, and 2) static and shared libs often require different compiler switches (`-fPIC` is needed for shared libs but not for static ones), so it might not make sense to allow both in a single build as it would mean compiling every object file twice anyway)

* some of the fancier switches that the autotools build supports (`-mfpmath=387` or link time optimization)

Things that I haven't tested and that I would need help with:

* Windows support. I still don't have a Windows machine so I have no idea how the CMake build would work there.

**Comment by ntamas** [comment]
And by the way, the CI tests also work [in Travis](https://travis-ci.org/github/igraph/igraph/builds/698636305). I'll start working on the Appveyor support only after someone has given me feedback about how the build works on Windows.

**Comment by vtraag** [comment]
Great to hear about the good progress @ntamas ! I could take a closer look on this Friday, I won't have much time before than.

**Comment by szhorvat** [comment]
Why is the build so much faster? :-)

For me, the blocker for testing on Windows is ARPACK. I don't *yet* know how to set that up (with MSVC), and that's probably going to take a while ...

**Comment by ntamas** [comment]
The CMake configuration is set up in a way that uses `ccache` for speeding up compilation if `ccache` is installed on your system. So the first build is slow, but subsequent builds are considerably faster for files that you have not touched, assuming you did not change the compiler switches either.

ARPACK is bundled and the build can use the bundled ARPACK (in theory).

**Comment by szhorvat** [comment]
I'm not familiar with cmake, except for building other packages that provide instructions, so I'm going to be slow ...

Now I see that it's `cmake .. -DIGRAPH_USE_INTERNAL_ARPACK=ON`

**Comment by ntamas** [comment]
On Linux and macOS, CMake has a console-based tool called `ccmake` that you can use to tweak the variables. So, I typically do something like:

```sh
mkdir build
cd build
cmake ..
ccmake .
make
```

The variables prefixed with `IGRAPH_` on the `ccmake` GUI are the ones that we should tweak.

**Comment by ntamas** [comment]
Progress report: the HTML and PDF documentation build now also seems to work, although I still need to work a bit on the PDF part so we can exclude the source codes of the examples from there.

One thing that is not ported over yet is the part that builds the documentation in GNU info format, and I'm not sure how much effort we should put into that. Is anyone using GNU info docs nowadays?

**Comment by szhorvat** [comment]
I vote to drop GNU info.

**Comment by vtraag** [comment]
> I vote to drop GNU info.

Agree, especially if it takes considerable time investment. Good HTML is more important I think.

**Comment by iosonofabio** [comment]
Yeah sure let's drop it

**Comment by ntamas** [comment]
Okay; in that case I guess there are three things left to be sorted out before we can switch to CMake:

* examples (i.e. the source codes of all the files in the `examples/` subdir) have to be stripped from the PDF output as they take up too much space

* we need to test whether the library builds properly if the tools that are required for the documentation only (DocBook, Apache FOP, `xsltproc`, `xmllint`, `source-highlight`) are not installed on the user's system

* sort out things for the Windows build. Right now the Appveyor builds fail because `bison` and `flex` are not installed, so I'd appreciate any input on how to do that nicely (someone mentioned `vcpkg` but I have no idea what that is :))

**Comment by vtraag** [comment]
>     * Right now the Appveyor builds fail because `bison` and `flex` are not installed, so I'd appreciate any input on how to do that nicely

I believe you can install it using `choco`, see https://chocolatey.org/packages/winflexbison.

**Comment by ntamas** [comment]
Closing this issue; it does not add much to the "CMake build system" milestone and all the pending tasks have separate issues there.
